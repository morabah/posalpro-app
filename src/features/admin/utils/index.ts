/**
 * PosalPro MVP2 - Admin Feature Utilities
 * Utility functions for admin feature operations
 * Based on CORE_REQUIREMENTS.md and ADMIN_MIGRATION_ASSESSMENT.md
 *
 * Component Traceability Matrix:
 * - User Stories: US-8.1, US-8.2
 * - Acceptance Criteria: AC-8.1.1, AC-8.1.2, AC-8.2.1, AC-8.2.2
 * - Hypotheses: H8
 * - Test Cases: TC-H8-001, TC-H8-002, TC-H8-003
 */

import type { User, Role, Permission, SystemMetrics, DatabaseStatus } from '../schemas';

/**
 * User status utilities
 */
export const userStatusUtils = {
  /**
   * Get user status color for UI display
   */
  getStatusColor: (status: User['status']): string => {
    switch (status) {
      case 'ACTIVE':
        return 'text-green-600 bg-green-50';
      case 'INACTIVE':
        return 'text-gray-600 bg-gray-50';
      case 'PENDING':
        return 'text-yellow-600 bg-yellow-50';
      case 'DELETED':
        return 'text-red-600 bg-red-50';
      case 'SUSPENDED':
        return 'text-orange-600 bg-orange-50';
      default:
        return 'text-gray-600 bg-gray-50';
    }
  },

  /**
   * Get user status badge variant
   */
  getStatusBadgeVariant: (status: User['status']): 'success' | 'warning' | 'error' | 'default' => {
    switch (status) {
      case 'ACTIVE':
        return 'success';
      case 'PENDING':
        return 'warning';
      case 'DELETED':
      case 'SUSPENDED':
        return 'error';
      default:
        return 'default';
    }
  },

  /**
   * Check if user status allows login
   */
  canLogin: (status: User['status']): boolean => {
    return status === 'ACTIVE';
  },

  /**
   * Get status display label
   */
  getStatusLabel: (status: User['status']): string => {
    return status.charAt(0).toUpperCase() + status.slice(1).toLowerCase();
  },
} as const;

/**
 * Role utilities
 */
export const roleUtils = {
  /**
   * Get role access level color
   */
  getAccessLevelColor: (level: Role['level']): string => {
    if (level >= 9) return 'text-red-600';
    if (level >= 7) return 'text-orange-600';
    if (level >= 5) return 'text-blue-600';
    if (level >= 3) return 'text-yellow-600';
    return 'text-gray-600';
  },

  /**
   * Get access level label
   */
  getAccessLevelLabel: (accessLevel: Role['accessLevel']): string => {
    return accessLevel;
  },

  /**
   * Check if role is system role
   */
  isSystemRole: (role: Role): boolean => {
    return role.isSystem === true;
  },

  /**
   * Get role hierarchy level
   */
  getHierarchyLevel: (role: Role): number => {
    return role.level;
  },

  /**
   * Sort roles by hierarchy level
   */
  sortByHierarchy: (roles: Role[]): Role[] => {
    return [...roles].sort((a, b) => b.level - a.level);
  },
} as const;

/**
 * Permission utilities
 */
export const permissionUtils = {
  /**
   * Get permission scope color
   */
  getScopeColor: (scope: Permission['scope']): string => {
    switch (scope) {
      case 'ALL':
        return 'text-red-600 bg-red-50';
      case 'TEAM':
        return 'text-blue-600 bg-blue-50';
      case 'OWN':
        return 'text-green-600 bg-green-50';
      default:
        return 'text-gray-600 bg-gray-50';
    }
  },

  /**
   * Format permission string for display
   */
  formatPermission: (resource: string, action: string): string => {
    return `${resource}:${action}`;
  },

  /**
   * Parse permission string
   */
  parsePermission: (permission: string): { resource: string; action: string } => {
    const [resource, action] = permission.split(':');
    return { resource: resource || '', action: action || '' };
  },

  /**
   * Group permissions by resource
   */
  groupByResource: (permissions: Permission[]): Record<string, Permission[]> => {
    return permissions.reduce((groups, permission) => {
      const resource = permission.resource;
      if (!groups[resource]) {
        groups[resource] = [];
      }
      groups[resource].push(permission);
      return groups;
    }, {} as Record<string, Permission[]>);
  },

  /**
   * Filter permissions by scope
   */
  filterByScope: (permissions: Permission[], scope: Permission['scope']): Permission[] => {
    return permissions.filter(permission => permission.scope === scope);
  },
} as const;

/**
 * System metrics utilities
 */
export const systemMetricsUtils = {
  /**
   * Get system health color
   */
  getHealthColor: (status: SystemMetrics['apiStatus'] | SystemMetrics['databaseStatus']): string => {
    const statusStr = status.toLowerCase();
    if (statusStr.includes('operational') || statusStr.includes('healthy')) {
      return 'text-green-600';
    }
    if (statusStr.includes('degraded') || statusStr.includes('warning')) {
      return 'text-yellow-600';
    }
    return 'text-red-600';
  },

  /**
   * Format storage size
   */
  formatStorageSize: (bytes: number): string => {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let size = bytes;
    let unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    return `${size.toFixed(1)} ${units[unitIndex]}`;
  },

  /**
   * Calculate storage percentage
   */
  calculateStoragePercentage: (used: number, total: number): number => {
    if (total === 0) return 0;
    return Math.round((used / total) * 100);
  },

  /**
   * Get performance indicator color
   */
  getPerformanceColor: (responseTime: number): string => {
    if (responseTime < 200) return 'text-green-600';
    if (responseTime < 500) return 'text-yellow-600';
    return 'text-red-600';
  },

  /**
   * Format uptime duration
   */
  formatUptime: (seconds: number): string => {
    const days = Math.floor(seconds / 86400);
    const hours = Math.floor((seconds % 86400) / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);

    if (days > 0) return `${days}d ${hours}h`;
    if (hours > 0) return `${hours}h ${minutes}m`;
    return `${minutes}m`;
  },
} as const;

/**
 * Database utilities
 */
export const databaseUtils = {
  /**
   * Get database health color
   */
  getHealthColor: (status: DatabaseStatus['health']): string => {
    switch (status) {
      case 'healthy':
        return 'text-green-600';
      case 'degraded':
        return 'text-yellow-600';
      case 'offline':
        return 'text-red-600';
      default:
        return 'text-gray-600';
    }
  },

  /**
   * Format latency for display
   */
  formatLatency: (latency: number | null): string => {
    if (latency === null) return 'N/A';
    if (latency < 1000) return `${latency}ms`;
    return `${(latency / 1000).toFixed(1)}s`;
  },

  /**
   * Get connection type label
   */
  getConnectionTypeLabel: (type: DatabaseStatus['connectionType']): string => {
    switch (type) {
      case 'local':
        return 'Local';
      case 'cloud':
        return 'Cloud';
      case 'hybrid':
        return 'Hybrid';
      default:
        return 'Unknown';
    }
  },

  /**
   * Check if database is online
   */
  isOnline: (status: DatabaseStatus): boolean => {
    return status.isOnline && status.health !== 'offline';
  },
} as const;

/**
 * Search and filter utilities
 */
export const searchUtils = {
  /**
   * Filter users by search term
   */
  filterUsers: (users: User[], searchTerm: string): User[] => {
    if (!searchTerm.trim()) return users;

    const term = searchTerm.toLowerCase();
    return users.filter(user =>
      user.name.toLowerCase().includes(term) ||
      user.email.toLowerCase().includes(term) ||
      user.department.toLowerCase().includes(term) ||
      user.role.toLowerCase().includes(term)
    );
  },

  /**
   * Filter roles by search term
   */
  filterRoles: (roles: Role[], searchTerm: string): Role[] => {
    if (!searchTerm.trim()) return roles;

    const term = searchTerm.toLowerCase();
    return roles.filter(role =>
      role.name.toLowerCase().includes(term) ||
      role.description.toLowerCase().includes(term)
    );
  },

  /**
   * Filter permissions by search term
   */
  filterPermissions: (permissions: Permission[], searchTerm: string): Permission[] => {
    if (!searchTerm.trim()) return permissions;

    const term = searchTerm.toLowerCase();
    return permissions.filter(permission =>
      permission.resource.toLowerCase().includes(term) ||
      permission.action.toLowerCase().includes(term) ||
      permission.displayName.toLowerCase().includes(term)
    );
  },
} as const;

/**
 * Validation utilities
 */
export const validationUtils = {
  /**
   * Validate email format
   */
  isValidEmail: (email: string): boolean => {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  },

  /**
   * Validate password strength
   */
  validatePassword: (password: string): { isValid: boolean; errors: string[] } => {
    const errors: string[] = [];

    if (password.length < 8) {
      errors.push('Password must be at least 8 characters long');
    }

    if (!/[A-Z]/.test(password)) {
      errors.push('Password must contain at least one uppercase letter');
    }

    if (!/[a-z]/.test(password)) {
      errors.push('Password must contain at least one lowercase letter');
    }

    if (!/\d/.test(password)) {
      errors.push('Password must contain at least one number');
    }

    if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password)) {
      errors.push('Password must contain at least one special character');
    }

    return {
      isValid: errors.length === 0,
      errors,
    };
  },

  /**
   * Validate role name
   */
  isValidRoleName: (name: string): boolean => {
    return name.trim().length >= 2 && name.trim().length <= 50;
  },

  /**
   * Validate permission format
   */
  isValidPermission: (permission: string): boolean => {
    const parts = permission.split(':');
    return parts.length === 2 && parts[0].trim().length > 0 && parts[1].trim().length > 0;
  },
} as const;

// Export all utilities as a single object for convenience
export const adminUtils = {
  userStatus: userStatusUtils,
  role: roleUtils,
  permission: permissionUtils,
  systemMetrics: systemMetricsUtils,
  database: databaseUtils,
  search: searchUtils,
  validation: validationUtils,
} as const;
