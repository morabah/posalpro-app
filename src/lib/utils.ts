/**
 * PosalPro MVP2 - Utility Functions
 * Common utility functions for the application
 */

import { type ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * Combine and merge Tailwind CSS classes
 * Handles conflicts and merges classes properly
 */
export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

/**
 * Debounce function to limit function calls
 */
export function debounce<A extends unknown[], R>(
  func: (...args: A) => R,
  wait: number
): (...args: A) => void {
  let timeout: ReturnType<typeof setTimeout>;

  return (...args: A) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      void func(...args);
    }, wait);
  };
}

/**
 * Throttle function to limit function calls
 */
export function throttle<A extends unknown[], R>(
  func: (...args: A) => R,
  limit: number
): (...args: A) => void {
  let inThrottle = false;

  return (...args: A) => {
    if (!inThrottle) {
      void func(...args);
      inThrottle = true;
      setTimeout(() => {
        inThrottle = false;
      }, limit);
    }
  };
}

/**
 * Format date to readable string
 */
export function formatDate(date: Date | string, options?: Intl.DateTimeFormatOptions): string {
  const dateObj = typeof date === 'string' ? new Date(date) : date;

  // Check if date is valid
  if (isNaN(dateObj.getTime())) {
    return 'Invalid Date';
  }

  try {
    return new Intl.DateTimeFormat('en-US', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      ...options,
    }).format(dateObj);
  } catch {
    return 'Invalid Date';
  }
}

/**
 * Format currency value
 */
export function formatCurrency(amount: number, currency = 'USD'): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
  }).format(amount);
}

/**
 * Format plain numbers with grouping and optional precision
 */
export function formatNumber(value: number, options: Intl.NumberFormatOptions = {}): string {
  const safeValue = Number.isFinite(value) ? value : 0;
  return new Intl.NumberFormat('en-US', {
    maximumFractionDigits: 2,
    ...options,
  }).format(safeValue);
}

/**
 * Format date in UTC with sensible defaults
 */
export function formatDateUTC(
  date: Date | string | number,
  options: Intl.DateTimeFormatOptions = {}
): string {
  const d = typeof date === 'string' || typeof date === 'number' ? new Date(date) : date;
  if (isNaN(d.getTime())) return 'Invalid Date';
  try {
    return new Intl.DateTimeFormat('en-US', {
      timeZone: 'UTC',
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      ...options,
    }).format(d);
  } catch {
    return 'Invalid Date';
  }
}

/**
 * Generate unique ID
 */
export function generateId(): string {
  return Math.random().toString(36).substring(2) + Date.now().toString(36);
}

/**
 * Sleep utility for async operations
 */
export function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Check if value is empty (null, undefined, empty string, empty array, empty object)
 */
export function isEmpty(value: unknown): boolean {
  if (value == null) return true;
  if (typeof value === 'string') return value.trim() === '';
  if (Array.isArray(value)) return value.length === 0;
  if (typeof value === 'object') return Object.keys(value as Record<string, unknown>).length === 0;
  return false;
}

/**
 * Safe JSON parse with fallback
 */
export function safeJsonParse<T>(str: string, fallback: T): T {
  try {
    return JSON.parse(str);
  } catch {
    return fallback;
  }
}

/**
 * Capitalize first letter of string
 */
export function capitalize(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Convert string to kebab-case
 */
export function toKebabCase(str: string): string {
  return str
    .replace(/([a-z])([A-Z])/g, '$1-$2')
    .replace(/\s+/g, '-')
    .toLowerCase();
}

/**
 * Convert string to camelCase
 */
export function toCamelCase(str: string): string {
  return str
    .replace(/(?:^\w|[A-Z]|\b\w)/g, (word, index) => {
      return index === 0 ? word.toLowerCase() : word.toUpperCase();
    })
    .replace(/\s+/g, '');
}

/**
 * Truncate string with ellipsis
 */
export function truncate(str: string, length: number): string {
  return str.length > length ? str.substring(0, length) + '...' : str;
}
