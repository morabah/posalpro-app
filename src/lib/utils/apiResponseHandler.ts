/**
 * API Response Handler Utilities
 * Defensive programming utilities for handling various API response structures
 * Prevents errors like "response.data.map is not a function"
 */

import { ErrorCodes } from '@/lib/errors/ErrorCodes';
import { ErrorHandlingService } from '@/lib/errors/ErrorHandlingService';

// Initialize error handling service
const errorHandlingService = ErrorHandlingService.getInstance();

// Type definitions for API responses
interface ApiResponse<T = unknown> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

interface PaginatedResponse<T = unknown> {
  data: T[];
  pagination?: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

interface ApiResponseWithPagination<T = unknown> extends ApiResponse<PaginatedResponse<T>> {}

/**
 * Safely extracts an array from an API response with unknown structure
 * Handles multiple common response patterns defensively
 */
export function extractArrayFromResponse<T = unknown>(
  response: ApiResponse<unknown>,
  arrayKey?: string,
  fallback: T[] = []
): T[] {
  if (!response) {
    // Use ErrorHandlingService for structured logging
    errorHandlingService.processError(
      new Error('No response provided to extractArrayFromResponse'),
      'API response extraction failed - no response',
      ErrorCodes.API.INVALID_RESPONSE,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        userFriendlyMessage: 'Invalid API response received.',
      }
    );
    return fallback;
  }

  // Log response structure for debugging with structured error handling
  errorHandlingService.processError(
    new Error('API Response structure analysis'),
    'Analyzing API response structure',
    ErrorCodes.SYSTEM.DEBUG_INFO,
    {
      component: 'apiResponseHandler',
      operation: 'extractArrayFromResponse',
      responseStructure: {
        success: response.success,
        hasData: !!response.data,
        dataType: typeof response.data,
        isDataArray: Array.isArray(response.data),
        keys: response.data ? Object.keys(response.data) : [],
      },
      userFriendlyMessage: 'Processing API response...',
    }
  );

  // Check if response indicates failure
  if (response.success === false) {
    errorHandlingService.processError(
      new Error('API response indicates failure'),
      'API operation failed',
      ErrorCodes.API.REQUEST_FAILED,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        responseError: response.error || response.message,
        userFriendlyMessage: 'The requested operation failed. Please try again.',
      }
    );
    return fallback;
  }

  // If no data property, return fallback
  if (!response.data) {
    errorHandlingService.processError(
      new Error('No data property in API response'),
      'Missing data in API response',
      ErrorCodes.API.INVALID_RESPONSE,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        userFriendlyMessage: 'No data returned from server.',
      }
    );
    return fallback;
  }

  const responseData = response.data as Record<string, unknown>;

  // Pattern 1: Direct array response
  if (Array.isArray(responseData)) {
    errorHandlingService.processError(
      new Error(`Found direct array with ${responseData.length} items`),
      'Direct array response detected',
      ErrorCodes.SYSTEM.DEBUG_INFO,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        pattern: 'direct_array',
        itemCount: responseData.length,
        userFriendlyMessage: `Found ${responseData.length} items.`,
      }
    );
    return responseData;
  }

  // Pattern 2: Specific key provided
  if (arrayKey && responseData[arrayKey] && Array.isArray(responseData[arrayKey])) {
    errorHandlingService.processError(
      new Error(`Found array in ${arrayKey} with ${responseData[arrayKey].length} items`),
      'Specific key array found',
      ErrorCodes.SYSTEM.DEBUG_INFO,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        pattern: 'specific_key',
        arrayKey,
        itemCount: responseData[arrayKey].length,
        userFriendlyMessage: `Found ${responseData[arrayKey].length} items.`,
      }
    );
    return responseData[arrayKey];
  }

  // Pattern 3: Common nested patterns
  const commonKeys = ['data', 'items', 'results', 'list', 'records'];
  for (const key of commonKeys) {
    if (responseData[key] && Array.isArray(responseData[key])) {
      errorHandlingService.processError(
        new Error(`Found array in ${key} with ${responseData[key].length} items`),
        'Common pattern array found',
        ErrorCodes.SYSTEM.DEBUG_INFO,
        {
          component: 'apiResponseHandler',
          operation: 'extractArrayFromResponse',
          pattern: 'common_nested',
          arrayKey: key,
          itemCount: responseData[key].length,
          userFriendlyMessage: `Found ${responseData[key].length} items.`,
        }
      );
      return responseData[key];
    }
  }

  // Pattern 4: Resource-specific patterns
  const resourceKeys = ['proposals', 'users', 'products', 'customers', 'teams'];
  for (const key of resourceKeys) {
    if (responseData[key] && Array.isArray(responseData[key])) {
      errorHandlingService.processError(
        new Error(`Found array in ${key} with ${responseData[key].length} items`),
        'Resource-specific array found',
        ErrorCodes.SYSTEM.DEBUG_INFO,
        {
          component: 'apiResponseHandler',
          operation: 'extractArrayFromResponse',
          pattern: 'resource_specific',
          arrayKey: key,
          itemCount: responseData[key].length,
          userFriendlyMessage: `Found ${responseData[key].length} items.`,
        }
      );
      return responseData[key];
    }
  }

  // Pattern 5: Search for any array property
  const arrayProperties = Object.entries(responseData)
    .filter(([_, value]) => Array.isArray(value))
    .map(([key, value]) => ({ key, array: value as T[] }));

  if (arrayProperties.length > 0) {
    const { key, array } = arrayProperties[0];
    errorHandlingService.processError(
      new Error(`Found array in property '${key}' with ${array.length} items`),
      'Dynamic array property found',
      ErrorCodes.SYSTEM.DEBUG_INFO,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        pattern: 'dynamic_search',
        arrayKey: key,
        itemCount: array.length,
        userFriendlyMessage: `Found ${array.length} items.`,
      }
    );
    return array;
  }

  // Pattern 6: Pagination structure
  if (responseData.pagination && responseData.data && Array.isArray(responseData.data)) {
    errorHandlingService.processError(
      new Error(`Found paginated array with ${responseData.data.length} items`),
      'Paginated array found',
      ErrorCodes.SYSTEM.DEBUG_INFO,
      {
        component: 'apiResponseHandler',
        operation: 'extractArrayFromResponse',
        pattern: 'paginated',
        itemCount: responseData.data.length,
        userFriendlyMessage: `Found ${responseData.data.length} items.`,
      }
    );
    return responseData.data;
  }

  // No array found - log comprehensive error
  errorHandlingService.processError(
    new Error('No array found in response structure'),
    'Array extraction failed',
    ErrorCodes.API.INVALID_RESPONSE,
    {
      component: 'apiResponseHandler',
      operation: 'extractArrayFromResponse',
      responseDataStructure: responseData,
      userFriendlyMessage: 'No data could be extracted from server response.',
    }
  );

  return fallback;
}

/**
 * Type-safe wrapper for PaginatedResponse extraction
 */
export function extractPaginatedArray<T = unknown>(
  response: ApiResponseWithPagination<T>,
  fallback: T[] = []
): { data: T[]; pagination?: PaginatedResponse<T>['pagination'] } {
  const data = extractArrayFromResponse<T>(response, undefined, fallback);
  const pagination = (response?.data as PaginatedResponse<T>)?.pagination;

  return {
    data,
    pagination,
  };
}

/**
 * Validates that a response has the expected structure
 */
export function validateResponseStructure(
  response: ApiResponse<unknown>,
  requiredFields: string[] = ['success']
): boolean {
  if (!response || typeof response !== 'object') {
    errorHandlingService.processError(
      new Error('Response is not an object'),
      'Invalid response structure',
      ErrorCodes.API.INVALID_RESPONSE,
      {
        component: 'apiResponseHandler',
        operation: 'validateResponseStructure',
        responseType: typeof response,
        userFriendlyMessage: 'Invalid server response format.',
      }
    );
    return false;
  }

  for (const field of requiredFields) {
    if (!(field in response)) {
      errorHandlingService.processError(
        new Error(`Missing required field '${field}'`),
        'Response validation failed',
        ErrorCodes.API.INVALID_RESPONSE,
        {
          component: 'apiResponseHandler',
          operation: 'validateResponseStructure',
          missingField: field,
          requiredFields,
          userFriendlyMessage: `Server response missing required field: ${field}`,
        }
      );
      return false;
    }
  }

  return true;
}

/**
 * Logs response structure for debugging
 */
export function debugResponseStructure(
  response: ApiResponse<unknown>,
  label = 'API Response'
): void {
  errorHandlingService.processError(
    new Error(`Debug response structure: ${label}`),
    'Response structure debug analysis',
    ErrorCodes.SYSTEM.DEBUG_INFO,
    {
      component: 'apiResponseHandler',
      operation: 'debugResponseStructure',
      label,
      responseAnalysis: {
        type: typeof response,
        keys: response ? Object.keys(response) : 'No response',
        dataType: response?.data ? typeof response.data : undefined,
        isDataArray: response?.data ? Array.isArray(response.data) : undefined,
        dataKeys:
          response?.data && typeof response.data === 'object' && !Array.isArray(response.data)
            ? Object.keys(response.data)
            : undefined,
      },
      fullResponse: response,
      userFriendlyMessage: `Debugging response structure for ${label}`,
    }
  );
}
