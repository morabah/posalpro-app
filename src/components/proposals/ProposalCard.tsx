/**
 * Proposal Card Component
 *
 * Displays a summary of a proposal with key metrics and status.
 * Follows our component composition patterns and design system.
 *
 * @stage Development
 * @quality-gate Component
 */

import React from 'react';
import { Badge } from '@/components/ui/Badge';
import { formatCurrency, formatDate } from '@/utils/formatters';

// Component Traceability Matrix
const COMPONENT_MAPPING = {
  userStories: ['US-3.1', 'US-3.2'],
  acceptanceCriteria: ['AC-3.1.1', 'AC-3.1.2', 'AC-3.2.1'],
  methods: ['displayProposalSummary()', 'calculateWinProbability()'],
  hypotheses: ['H2'],
  testCases: ['TC-H2-001', 'TC-H2-002'],
};

// Type definitions for component props
export interface ProposalCardProps {
  /** Unique identifier for the proposal */
  id: string;
  /** Title of the proposal */
  title: string;
  /** Current status of the proposal */
  status: 'draft' | 'in-review' | 'submitted' | 'won' | 'lost';
  /** Client or organization name */
  client: string;
  /** Total proposal value */
  value: number;
  /** Probability of winning (0-100) */
  winProbability?: number;
  /** Due date for the proposal */
  dueDate: string | Date;
  /** Last modified timestamp */
  lastModified: string | Date;
  /** Optional click handler */
  onClick?: (id: string) => void;
  /** Optional additional class names */
  className?: string;
}

/**
 * Maps proposal status to badge variant for visual indication
 */
const getStatusBadgeVariant = (status: ProposalCardProps['status']) => {
  const variantMap: Record<
    ProposalCardProps['status'],
    React.ComponentProps<typeof Badge>['variant']
  > = {
    draft: 'secondary',
    'in-review': 'warning',
    submitted: 'default',
    won: 'success',
    lost: 'destructive',
  };

  return variantMap[status] || 'default';
};

/**
 * Formats status for display with proper capitalization and hyphen removal
 */
const formatStatus = (status: string): string => {
  return status
    .split('-')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
};

/**
 * ProposalCard component displays a summary card for a proposal
 */
export const ProposalCard: React.FC<ProposalCardProps> = ({
  id,
  title,
  status,
  client,
  value,
  winProbability = 0,
  dueDate,
  lastModified,
  onClick,
  className = '',
}) => {
  const handleClick = () => {
    if (onClick) {
      onClick(id);
    }
  };

  return (
    <div
      className={`flex flex-col p-4 bg-white rounded-lg shadow-sm border border-gray-200 hover:shadow-md transition-shadow ${className}`}
      onClick={handleClick}
      data-testid="proposal-card"
    >
      <div className="flex justify-between items-start mb-2">
        <h3 className="text-lg font-semibold text-gray-900 truncate" title={title}>
          {title}
        </h3>
        <Badge variant={getStatusBadgeVariant(status)}>{formatStatus(status)}</Badge>
      </div>

      <p className="text-sm text-gray-600 mb-3">{client}</p>

      <div className="grid grid-cols-2 gap-2 mb-3">
        <div className="flex flex-col">
          <span className="text-xs text-gray-500">Value</span>
          <span className="text-sm font-medium">{formatCurrency(value)}</span>
        </div>

        <div className="flex flex-col">
          <span className="text-xs text-gray-500">Win Probability</span>
          <span className="text-sm font-medium">{winProbability}%</span>
        </div>
      </div>

      <div className="grid grid-cols-2 gap-2 mt-auto pt-2 border-t border-gray-100">
        <div className="flex flex-col">
          <span className="text-xs text-gray-500">Due Date</span>
          <span className="text-xs">{formatDate(dueDate, 'short')}</span>
        </div>

        <div className="flex flex-col">
          <span className="text-xs text-gray-500">Last Modified</span>
          <span className="text-xs">{formatDate(lastModified, 'short')}</span>
        </div>
      </div>
    </div>
  );
};
