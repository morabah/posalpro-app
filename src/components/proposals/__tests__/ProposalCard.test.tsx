/**
 * ProposalCard Component Tests
 *
 * Includes snapshot testing and interaction testing for ProposalCard component.
 * Follows our quality-first approach and testing best practices.
 *
 * @stage Development
 * @quality-gate Component Testing
 */

import React from 'react';
import { render, screen, fireEvent, cleanup } from '@/test/utils/test-utils';
import { ProposalCard } from '../ProposalCard';

describe('ProposalCard Component', () => {
  const mockProposal = {
    id: 'proposal-123',
    title: 'Enterprise Cloud Migration Proposal',
    status: 'in-review' as const,
    client: 'Acme Corporation',
    value: 150000,
    winProbability: 75,
    dueDate: '2025-07-15T00:00:00.000Z',
    lastModified: '2025-06-01T14:30:00.000Z',
  };

  it('renders correctly with all props', () => {
    const { container } = render(<ProposalCard {...mockProposal} />);

    // Verify key elements are present
    expect(screen.getByText('Enterprise Cloud Migration Proposal')).toBeInTheDocument();
    expect(screen.getByText('Acme Corporation')).toBeInTheDocument();
    expect(screen.getByText('In Review')).toBeInTheDocument();
    expect(screen.getByText('75%')).toBeInTheDocument();

    // Take a snapshot
    expect(container).toMatchSnapshot();
  });

  it('renders with correct status badge variant', () => {
    // Test for each status type
    const statuses = ['draft', 'in-review', 'submitted', 'won', 'lost'] as const;

    statuses.forEach(status => {
      render(<ProposalCard {...mockProposal} status={status} />);

      // Check if status text is rendered with proper formatting
      const expectedText =
        status === 'in-review' ? 'In Review' : status.charAt(0).toUpperCase() + status.slice(1);
      const statusBadge = screen.getByText(expectedText);
      expect(statusBadge).toBeInTheDocument();

      // Clean up for next iteration
      cleanup();
    });
  });

  it('calls onClick handler when clicked', () => {
    const handleClick = jest.fn();
    render(<ProposalCard {...mockProposal} onClick={handleClick} />);

    const card = screen.getByTestId('proposal-card');
    fireEvent.click(card);

    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(handleClick).toHaveBeenCalledWith(mockProposal.id);
  });

  it('formats currency and dates correctly', () => {
    render(<ProposalCard {...mockProposal} />);

    // Check currency formatting
    expect(screen.getByText('$150,000.00')).toBeInTheDocument();

    // Check date formatting (depends on locale settings in formatters)
    // This assumes the short date format, which should be predictable
    expect(screen.getByText(/7\/15\/2025/)).toBeInTheDocument();
    expect(screen.getByText(/6\/1\/2025/)).toBeInTheDocument();
  });

  it('renders without crashing when winProbability is not provided', () => {
    const { winProbability, ...proposalWithoutWinProb } = mockProposal;
    render(<ProposalCard {...proposalWithoutWinProb} />);

    expect(screen.getByText('0%')).toBeInTheDocument();
  });
});
