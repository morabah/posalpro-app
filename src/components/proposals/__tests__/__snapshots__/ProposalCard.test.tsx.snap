// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalCard Component renders correctly with all props 1`] = `
<div>
  <div
    data-testid="mock-auth-provider"
  >
    <div
      class="flex flex-col p-4 bg-white rounded-lg shadow-sm border border-gray-200 hover:shadow-md transition-shadow "
      data-testid="proposal-card"
    >
      <div
        class="flex justify-between items-start mb-2"
      >
        <h3
          class="text-lg font-semibold text-gray-900 truncate"
          title="Enterprise Cloud Migration Proposal"
        >
          Enterprise Cloud Migration Proposal
        </h3>
        <span
          class="inline-flex items-center rounded-full font-medium bg-yellow-100 text-yellow-800 px-2.5 py-0.5 text-sm"
        >
          In Review
        </span>
      </div>
      <p
        class="text-sm text-gray-600 mb-3"
      >
        Acme Corporation
      </p>
      <div
        class="grid grid-cols-2 gap-2 mb-3"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs text-gray-500"
          >
            Value
          </span>
          <span
            class="text-sm font-medium"
          >
            $150,000.00
          </span>
        </div>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs text-gray-500"
          >
            Win Probability
          </span>
          <span
            class="text-sm font-medium"
          >
            75
            %
          </span>
        </div>
      </div>
      <div
        class="grid grid-cols-2 gap-2 mt-auto pt-2 border-t border-gray-100"
      >
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs text-gray-500"
          >
            Due Date
          </span>
          <span
            class="text-xs"
          >
            7/15/2025
          </span>
        </div>
        <div
          class="flex flex-col"
        >
          <span
            class="text-xs text-gray-500"
          >
            Last Modified
          </span>
          <span
            class="text-xs"
          >
            6/1/2025
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
